src/main/scala/com/fulcrumgenomics/vcf/api/Variant.scala

```diff
@@ -0,0 +1,183 @@
+
+import scala.collection.immutable.ListMap
+
+object Variant {
+  /** Value used in VCF for values that are missing. */
+  val Missing: String = "."
+
+  /** The value stored for fields of type `Flag` when stored in a Map. */
+  val FlagValue: String = Missing
+
+  /** The type used in attribute maps (INFO, genotype) to store multi-valued attributed. */
+  type ArrayAttribute[A] = IndexedSeq[A]
+
+  /** Function to convert any scala collection to an [[com.fulcrumgenomics.vcf.api.Variant.ArrayAttribute]]. */
+  def toArrayAttribute[A](values: TraversableOnce[A]): ArrayAttribute[A] = values.toIndexedSeq
+
+  /** The set of filters that are applied to passing variants. */
+  val PassingFilters: Set[String] = Set("PASS")
+}
+
+
+/**
+  * Represents a variant from a VCF or similar source.
+  *
+  * @param chrom the chromosome on which a variant resides.
+  * @param pos the start position of the variant
+  * @param id the optional ID (e.g. rs number) of the variant
+  * @param alleles the set of alleles recorded for the variant
+  * @param qual the optional phred-scaled quality score of the variant
+  * @param filter the set of filters applied to the variant.  An empty set indicates the variant has not had
+  *               filtration applied.  A single value of "PASS" is applied if the variant passes filters, and
+  *               one or more non-PASS strings if the variant fails filters.
+  * @param attributes a map of attributes that come from the INFO field of a VCF
+  * @param genotypes a map of sample name -> genotype for the variant
+  */
+case class Variant(chrom: String,
+                   pos: Int,
+                   id: Option[String],
+                   alleles: AlleleSet,
+                   qual: Option[Double],
+                   filter: Set[String],
+                   attributes: ListMap[String,Any],
+                   genotypes: Map[String, Genotype]
+                  ) {
+
+  /** The end position of the variant based on either the `END` INFO field _or_ the length of the reference allele. */
+  val end: Int = get[Int]("END").getOrElse(pos + alleles.ref.length - 1)
+
+  /** Retrieves a value from the INFO map.  Will throw an exception if the key does not exist. */
+  def apply[A](key: String): A = attributes(key).asInstanceOf[A]
+
+  /** Retrieves an optional value from the INFO map.  Will return [[None]] if the key does not exist. */
+  def get[A](key: String): Option[A] = attributes.get(key).asInstanceOf[Option[A]]
+
+  /** Retrieves an optional value from the INFO map.  Will return `default` if the key does not exist. */
+  def getOrElse[A](key: String, default: => A): Option[A] = attributes.getOrElse(key, default).asInstanceOf[Option[A]]
+
+  /** Object that provides convenient accessors to common, spec-defined, attributes from the INFO field.
+    * Methods on this object return the expected type directly and will throw exceptions if the keys
+    * do not exist in the info map.
+    */
+  final object INFO {
```

Comment:
@nh13 So this seems to work.  I'm having a hard time coming up with names for the pair of objects here and again in `Genotype`, so suggestions are welcome.

I renamed `info` to `attributes` in the constructor for a couple of reasons:
1) So I could use `info` here
2) So it was less obvious when using the object; it does need to be exposed though otherwise it makes calling the copy() methods really awkward